src/front_landview.c

```diff
@@ -1371,43 +1373,115 @@ void draw_netmap_players_hands(void)
 }
 
 /**
- * Draws text description of active level.
+ * Returns order number (starting at 1) for Campaign->bonus_levels associated with given bonus level.
+ * If the level is not found, returns -1.
  */
-void draw_map_level_descriptions(void)
+int order_number_for_bonus_level(LevelNumber bn_lvnum)
 {
-  if ((fe_net_level_selected > 0) || (net_level_hilighted > 0))
+  int orderNum = 1;
+  if (bn_lvnum < 1) return -1;
+  for (int i = 0; i < CAMPAIGN_LEVELS_COUNT; i++)
   {
-    lbDisplay.DrawFlags = 0;
-    LevelNumber lvnum = fe_net_level_selected;
-    if (lvnum <= 0)
-      lvnum = net_level_hilighted;
-    struct LevelInformation* lvinfo = get_level_info(lvnum);
-    if (lvinfo == NULL)
-      return;
-    const char* lv_name;
-    if (lvinfo->name_stridx > 0)
-        lv_name = get_string(lvinfo->name_stridx);
-    else
-      lv_name = lvinfo->name;
-    if ((lv_name != NULL) && (strlen(lv_name) > 0)) {
-      snprintf(level_name, sizeof(level_name), "%s %d: %s", get_string(GUIStr_MnuLevel), (int)lvinfo->lvnum, lv_name);
-    } else {
-      snprintf(level_name, sizeof(level_name), "%s %d", get_string(GUIStr_MnuLevel), (int)lvinfo->lvnum);
-    }
-    long w = LbTextStringWidth(level_name);
-    long x = lvinfo->ensign_x - map_info.screen_shift_x;
-    long y = lvinfo->ensign_y - map_info.screen_shift_y - 8;
-    long h = LbTextHeight(level_name);
-    LbDrawBox(scale_value_landview(x-4), scale_value_landview(y), scale_value_landview(w+8), scale_value_landview(h), 0);
-    LbTextDrawResized(scale_value_landview(x), scale_value_landview(y), units_per_pixel_landview, level_name);
+    if (campaign.bonus_levels[i] == bn_lvnum)
+    {
+      return orderNum;
+    }
+    else if (campaign.bonus_levels[i] != 0)
+    {
+      orderNum++;
+    }
+  }
+  return -1;
+}
+
+/**
+ * Get text description of level.
+ */
+const char* get_level_description(struct LevelInformation *lvinfo)
+{
+  if (lvinfo == NULL)
+  {
+    return NULL;
+  }
+
+  if (lvinfo->options & LvOp_IsSingle || lvinfo->options & LvOp_IsMulti|| lvinfo->options & LvOp_Tutorial)
+  {
+      if (lvinfo->name_stridx > 0)
+      {
+        return get_string(lvinfo->name_stridx);
+      }
+      else
+      {
+        return lvinfo->name;
+      }
+  }
+
+  if (lvinfo->options & LvOp_IsBonus)
+  {
+    static char name_and_num[32];
+    snprintf(name_and_num, sizeof(name_and_num), "%s %d", get_string(CpgStr_BonusLevel), (int)order_number_for_bonus_level(lvinfo->lvnum));
+    return name_and_num;
+  }
+
+  if (lvinfo->options & LvOp_IsExtra)
+  {
+    // TODO: Implement translation / get_string
```

Comment:
Full Moon, New Moon, I think the best thing to do here is to just go with 'Bonus' too.